spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -589,6 +589,24 @@ private Long getStartSnapshotId(Long startTimestamp) {
     }
   }
 
+  private Long getEndSnapshotId(Long endTimestamp) {
+    Long snapshotId = null;
+    Snapshot current = table.currentSnapshot();
+    if (current != null) {
+      if (current.timestampMillis() <= endTimestamp) {
+        snapshotId = current.snapshotId();
+      } else {
+        for (Snapshot ancestor : SnapshotUtil.currentAncestors(table)) {
+          if (ancestor.timestampMillis() <= endTimestamp) {
+            snapshotId = ancestor.snapshotId();
+            break;
+          }
+        }
+      }
+    }
```

Comment:
`SnapshotUtil.currentAncestors(table)` includes the current snapshot as well. We could simplify the logic a bit.

Reply:
Also I think we can move this method to the class `SnapshotUtil`. It could be useful for other scans as well. For example, I'm not sure if time travel query like this has the similar bug. We can double check on that, but it's a blocker for this PR.
```
SELECT * FROM prod.db.table TIMESTAMP AS OF '1986-10-26 01:21:00';
```